File: 02138. Divide a String Into Groups of Size k/solution.js
/**
 * Problem: 2138. Divide a String Into Groups of Size k
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Divides a string into groups of k characters, padding with fill character if needed
 *
 * @param {string} s - Input string to divide
 * @param {number} k - Characters per group
 * @param {string} fill - Padding character
 *
 * @returns {string[]} - Array of k-character strings
 */
const divideString = (s, k, fill) => {
  // Create a working copy of the input string
  let paddedString = s

  // Pad the string with fill character until length is divisible by k
  while (paddedString.length % k !== 0) paddedString += fill

  // Initialize result array with the correct size
  const result = new Array(paddedString.length / k)

  // Split the padded string into groups of k characters
  for (let i = 0; i < paddedString.length / k; i++)
    result[i] = paddedString.substring(i * k, (i + 1) * k)

  return result
}

File: 02138. Divide a String Into Groups of Size k/solution.go
/**
 * Problem: 2138. Divide a String Into Groups of Size k
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func divideString(s string, k int, fill byte) []string {
	// Initialize slice to store the result groups
	groups := []string{}
	// Get the length of input string
	length := len(s)

	for i := 0; i < length; i += k {
		// Calculate the end index for current chunk
		endIndex := i + k

		// Handle case where remaining characters are less than k
		if endIndex > length {
			// Extract remaining characters
			chunk := s[i:]
			// Create padding with fill character
			padding := make([]byte, k-len(chunk))

			// Fill padding with the specified character
			for j := range padding {
				padding[j] = fill
			}

			// Append padded chunk to result
			chunk += string(padding)
			groups = append(groups, chunk)
		} else {
			// Handle normal case where we have exactly k characters
			groups = append(groups, s[i:endIndex])
		}
	}

	return groups
}
